daemon/list.go

```diff
@@ -103,55 +106,72 @@ func (daemon *Daemon) Containers(ctx context.Context, config *containertypes.Lis
 		return nil, err
 	}
 
-	var (
-		view       = daemon.containersReplica.Snapshot()
-		containers = []*containertypes.Summary{}
-	)
-
+	view := daemon.containersReplica.Snapshot()
 	filter, err := daemon.foldFilter(ctx, view, config)
 	if err != nil {
 		return nil, err
 	}
 
-	// fastpath to only look at a subset of containers if specific name
+	// shortcut to only look at a subset of containers if specific name
 	// or ID matches were provided by the user--otherwise we potentially
 	// end up querying many more containers than intended
+	//
+	// TODO (thaJeztah): given containersReplica.Snapshot() provides a "consistent read-only view of the database" (which indicates "de-referenced copy", is there any reason we wouldn't use a []*container.Snapshot (pointer-slice)?.
 	containerList, err := daemon.filterByNameIDMatches(view, filter)
 	if err != nil {
 		return nil, err
 	}
+	numContainers := len(containerList)
+
+	// Get the info for each container in the list; this can be slow so we
+	// dispatch a set number of worker goroutines to do the jobs. We choose
+	// log2(numContainers) workers to avoid creating too many goroutines
+	// for large number of containers.
+	numWorkers := int(math.Log2(float64(numContainers)))
+	if numWorkers < 1 {
+		numWorkers = 1
+	}
 
-	for i := range containerList {
-		currentContainer := &containerList[i]
-		switch includeContainerInList(currentContainer, filter) {
+	resultsMut := sync.Mutex{}
+	results := make([]*containertypes.Summary, 0, min(len(containerList), filter.Limit))
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(numWorkers)
+
+	for _, ctr := range containerList {
+		switch includeContainerInList(&ctr, filter) {
 		case excludeContainer:
 			continue
 		case stopIteration:
-			return containers, nil
-		}
+			break
+		default:
+			filter.idx++
```

Comment:
```suggestion
			idx := idx
			filter.idx++
```

Reply:
Where does your `idx` come from here? (curious here because `len(containerList)` may be larger than `filter.Limit` here, as the `excludeContainers` branch would skip a container. (i.e., wondering if that works with `results` having a fixed size.

There may be other cases where we're getting less results than requested (filtering may leave out containers and we don't reach `filter.Limit`); would the results end up having empty entries at the end?


If those are still "to be looked into", we could still opt for "just sort the results at the end" (for now), and leave these for a follow-up improvement

Reply:
It's a zero value, but if we're concerned, we can just create our own `idx := 0` before the loop

Reply:
Yes containerList can be larger, but we can just limit it to the last `idx` (which is only incremented for non-skipped containers).
See https://github.com/moby/moby/pull/49365#discussion_r1942544299

Reply:
Ah! Gotcha; sometimes looking at review comments is like peeking through a letterbox, and you need to look at the things within full context of the code 😂 ❤️

Reply:
I opted for `idx = filter.idx` which is the same thing, just thought it was a bit clearer.